File: node_agent/src/snapshotter.js
/**
 * Snapshotter for Node.js - captures function arguments and local state
 */
class Snapshotter {
    constructor(config = {}) {
        this.config = {
            params: config.params !== false,
            fields: config.fields || [],
            locals: config.locals !== false,
            maxDepth: config.maxDepth || 5,
            maxProps: config.maxProps || 50,
        };
        this.visitedObjects = new Set();
    }

    /**
     * Capture snapshot of arguments and context
     */
    captureSnapshot(thisObj, args, locals = {}) {
        this.visitedObjects.clear();
        const snapshot = {};

        if (this.config.params && args) {
            for (let i = 0; i < args.length; i++) {
                snapshot[`arg${i}`] = this.safeSerialize(args[i], 0);
            }
        }

        if (thisObj && this.config.locals) {
            snapshot.this = this.safeSerialize(thisObj, 0);
        }

        if (locals && this.config.locals) {
            for (const [key, value] of Object.entries(locals)) {
                snapshot[key] = this.safeSerialize(value, 0);
            }
        }

        return snapshot;
    }

    /**
     * Safely serialize a value with depth and property limits
     */
    safeSerialize(obj, depth) {
        if (obj === null || obj === undefined) {
            return null;
        }

        if (depth >= this.config.maxDepth) {
            return '[depth limit reached]';
        }

        // Primitives and strings
        if (typeof obj === 'string' || typeof obj === 'number' ||
            typeof obj === 'boolean' || typeof obj === 'symbol') {
            return String(obj);
        }

        // Handle cycles
        const objId = this._getObjectId(obj);
        if (this.visitedObjects.has(objId)) {
            return '[cyclic reference]';
        }
        this.visitedObjects.add(objId);

        try {
            if (Array.isArray(obj)) {
                return this._serializeArray(obj, depth);
            } else if (obj instanceof Date) {
                return obj.toISOString();
            } else if (obj instanceof RegExp) {
                return obj.toString();
            } else if (obj instanceof Error) {
                return { message: obj.message, stack: obj.stack };
            } else if (typeof obj === 'object') {
                return this._serializeObject(obj, depth);
            }
            return String(obj);
        } finally {
            this.visitedObjects.delete(objId);
        }
    }

    _serializeArray(arr, depth) {
        const result = [];
        let count = 0;
        for (const item of arr) {
            if (count >= this.config.maxProps) {
                result.push({ __truncated__: true });
                break;
            }
            result.push(this.safeSerialize(item, depth + 1));
            count++;
        }
        return result;
    }

    _serializeObject(obj, depth) {
        const result = { __class__: obj.constructor.name };
        let count = 0;

        for (const [key, value] of Object.entries(obj)) {
            if (count >= this.config.maxProps) {
                result.__truncated__ = true;
                break;
            }
            if (!this.config.fields || this.config.fields.length === 0 ||
                this.config.fields.includes(key)) {
                result[key] = this.safeSerialize(value, depth + 1);
                count++;
            }
        }

        return result;
    }

    _getObjectId(obj) {
        // Use JSON serialization as a simple way to detect cycles
        // In production, would use WeakMap
        if (!obj.__debugin_id__) {
            obj.__debugin_id__ = Math.random().toString(36);
        }
        return obj.__debugin_id__;
    }
}

module.exports = Snapshotter;

File: agent-core/src/main/java/com/debugin/snapshot/Snapshotter.java
package com.debugin.snapshot;

import com.debugin.probe.JavaProbe;
import java.util.*;
import java.util.logging.Logger;

/**
 * Captures snapshots of method parameters, fields, and local variables.
 * Enforces depth and property limits while detecting cycles.
 */
public class Snapshotter {
    private static final Logger logger = Logger.getLogger(Snapshotter.class.getName());

    private final JavaProbe.SnapshotConfig config;
    private final Set<Integer> visitedObjects = new HashSet<>();

    public Snapshotter(JavaProbe.SnapshotConfig config) {
        this.config = config != null ? config : new JavaProbe.SnapshotConfig();
    }

    /**
     * Capture a snapshot of arguments, this, and locals
     */
    public Map<String, Object> captureSnapshot(Object thiz, Object[] args, Map<String, Object> locals) {
        visitedObjects.clear();
        Map<String, Object> snapshot = new LinkedHashMap<>();

        if (config.params && args != null) {
            for (int i = 0; i < args.length; i++) {
                snapshot.put("arg" + i, safeSerialize(args[i], 0));
            }
        }

        if (thiz != null && config.locals) {
            snapshot.put("this", safeSerialize(thiz, 0));
        }

        if (locals != null) {
            for (Map.Entry<String, Object> entry : locals.entrySet()) {
                snapshot.put(entry.getKey(), safeSerialize(entry.getValue(), 0));
            }
        }

        return snapshot;
    }

    /**
     * Safely serialize an object, respecting depth and property limits
     */
    private Object safeSerialize(Object obj, int depth) {
        if (obj == null) {
            return null;
        }

        // Check depth limit
        if (depth >= config.maxDepth) {
            return "[depth limit reached]";
        }

        // Detect cycles
        int objHash = System.identityHashCode(obj);
        if (visitedObjects.contains(objHash)) {
            return "[cyclic reference]";
        }

        // Handle primitives and strings
        if (obj instanceof String || obj instanceof Number || obj instanceof Boolean) {
            return obj;
        }

        visitedObjects.add(objHash);

        try {
            if (obj instanceof Map) {
                return serializeMap((Map<?, ?>) obj, depth);
            } else if (obj instanceof Collection) {
                return serializeCollection((Collection<?>) obj, depth);
            } else if (obj instanceof Object[]) {
                return serializeArray((Object[]) obj, depth);
            } else {
                return serializeObject(obj, depth);
            }
        } finally {
            visitedObjects.remove(objHash);
        }
    }

    private Object serializeMap(Map<?, ?> map, int depth) {
        Map<String, Object> result = new LinkedHashMap<>();
        int count = 0;

        for (Map.Entry<?, ?> entry : map.entrySet()) {
            if (count >= config.maxProps) {
                result.put("__truncated__", true);
                break;
            }
            String key = String.valueOf(entry.getKey());
            result.put(key, safeSerialize(entry.getValue(), depth + 1));
            count++;
        }

        return result;
    }

    private Object serializeCollection(Collection<?> coll, int depth) {
        List<Object> result = new ArrayList<>();
        int count = 0;

        for (Object item : coll) {
            if (count >= config.maxProps) {
                result.add(Collections.singletonMap("__truncated__", true));
                break;
            }
            result.add(safeSerialize(item, depth + 1));
            count++;
        }

        return result;
    }

    private Object serializeArray(Object[] arr, int depth) {
        List<Object> result = new ArrayList<>();
        int count = 0;

        for (Object item : arr) {
            if (count >= config.maxProps) {
                result.add(Collections.singletonMap("__truncated__", true));
                break;
            }
            result.add(safeSerialize(item, depth + 1));
            count++;
        }

        return result;
    }

    private Object serializeObject(Object obj, int depth) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("__class__", obj.getClass().getName());

        // Try to get fields using reflection
        try {
            Class<?> clazz = obj.getClass();
            java.lang.reflect.Field[] fields = clazz.getDeclaredFields();
            int count = 0;

            for (java.lang.reflect.Field field : fields) {
                if (count >= config.maxProps) {
                    result.put("__truncated__", true);
                    break;
                }

                // Skip if field is not in snapshot config
                if (!config.fields.isEmpty() && !config.fields.contains(field.getName())) {
                    continue;
                }

                field.setAccessible(true);
                try {
                    Object value = field.get(obj);
                    result.put(field.getName(), safeSerialize(value, depth + 1));
                    count++;
                } catch (IllegalAccessException e) {
                    result.put(field.getName(), "[access denied]");
                }
            }
        } catch (Exception e) {
            result.put("__error__", e.getMessage());
        }

        return result;
    }
}
